Bugfix for logging on a target even after removed #2227

jojosardez opened the pull request:
* Fixed the logging configuration to also remove the target from the configItems.
* Fixed the log factory to skip setting targets by level if the target doesn't exists (removed). 
* Added unit test (from the issue) to test the behavior.

304NotModified commented:
Thanks! I'm currently without internet on my PC, will try to review next week

jojosardez commented:
Thanks @304NotModified! By the way, i'm not sure what's wrong with AppVeyor since the same test runs successfully in Travis.

snakefoot commented:
@jojosardez Would be better to remove the target from the LoggingRules, instead of checking Configuration.AllTargets. The check is also a little wrong incase a LogginRule has multiple targets, then it will `break;` when meeting an unknown target.

jojosardez commented:
Hi @snakefoot , thanks for the comments! I'll look into it this weekend.

jojosardez commented:
Hi @snakefoot , your suggestion works. Added the removal of the target from the logging rule and removed the condition and break from in the loop in log factory.

AppVeyor still fails though even if the build is successful in Travis. The problem is unrelated to the changes made. FYI @304NotModified .

snakefoot commented:
@jojosardez Looks like the LoggingRules should be used like this:

```
var loggingRules = this.LoggingRules.ToList();  // Make clone to minimize concurrency issues
foreach (var rule in loggingRules)
{
    // Blah blah blah
}
```

snakefoot commented:
@jojosardez And the configItems should be updated like this:

```
var roots = new List<object>();
roots.AddRange(configItems);
roots.AddRange(this.targets);
this.configItems = ObjectGraphScanner.FindReachableObjects<object>(roots.ToArray());
```

Maybe consider to "fix" the existing method ValidateConfig() so it can be re-used for updating `this.configItems`.

snakefoot commented:
@jojosardez It seems you cannot use RemoveAll() for removing the target.

snakefoot commented:
@jojosardez One might also consider to Flush/Close/Dispose the target when removing it.

snakefoot commented:
@jojosardez When removing a target, then one could also consider to iterate through all wrapper-targets, and check if one is pointing to this target, and then remove it.

But if the wrapper-target is able to reference multiple targets, then it will suddenly be very difficult to remove the target (CompoundTargetBase).

Maybe the best thing is just to close/dispose the target (so it is no longer is initialized), and remove it from loggingrules. Anything more will just generate too much headache. With the small catch that if the removed target is a wrapper-target then it should also close/dispose the inner-target (unless it is being used by other logging-rules).

jojosardez commented:
Hi @snakefoot , sorry I've been very busy this past week. Anyway, I have changed the implementation keeping in mind your suggestions. I ditched the remove all and just remove the targets individually. I also reused ValidateConfig() instead of removing the configItems explicitly. I also added closing of the target and a unit test to check if it really happened, before their removal.

snakefoot commented:
@jojosardez Multiple loggingRules can include the same target. Instead of closing the target just after having removed it from the loggingRule, then you should delay the closing of target until having removed it from all loggingRules.

Maybe the method should look like this:

```
public void RemoveTarget(string name)
{
    HashSet<Target> removedTargets = new HashSet<Target>();
    Target removedTarget;
    if (this.targets.TryGetValue(name, out removedTarget))
    {
        removedTargets.Add(removedTarget);
        this.targets.Remove(name);
    }

    var loggingRules = LoggingRules.ToList();
    foreach (var rule in loggingRules)
    {
        var targetList = rule.Targets.ToList();
        foreach (var target in targetList)
        {
            if (target.Name == name)
            {
                removedTargets.Add(removedTarget);
                this.targets.Remove(name);
            }
        }
    }

    foreach (var target in removedTargets)
    {
        target.Close(); 
    }
    
    ValidateConfig();
}
```

304NotModified commented:
> @jojosardez Multiple loggingRules can include the same target.

good point, rules and targets has more on more relationships. Could we make an unit test for this case?

304NotModified commented:
This need some changes, right?

304NotModified commented:
please open a new PR when you have time, thanks!

304NotModified commented:
merged by #2549
